### DISCUSSION TITLE
Communication between turn server and pion via private ip with pion also accessible via public ip

### DISCUSSION THREAD

**Discussion by mysolo** [pion/webrtc#2421] [Q&A] [answered]
Hello,

We use pion on our encoders using SetNAT1To1IPs with ICECandidateTypeHost. This works perfectly when pion is directly accessible from the internet with firefox and chrome browsers. 
But we need to set up a turn server (coturn) to answer some access problems.
This one is on the same private network and we need it to communicate with pion on the private network.
We also have janus which must also be able to communicate with turn on the private network.

So we have configured Coturn in mode:
--external-ip=ip public/ip private

Janus has been configured : 
```
[nat]
nat_1_1_mapping = ip public,ip private
keep_private_host = true
```

With janus it works with chrome and firefox if you put the private ip in the nat mapping and with the keep_private_host 

With pion, the ICECandidate Type Host is not usable with private ip. So we used a Srflx type, and indicated the public and private ip.
It works under chrome but it remains blocked under firefox.

So we need to be able to send candidates by pion in the same way as janus of the type:
```
a=candidate:1 1 udp 2013266431 ip private 10101 typ host
a=candidate:1 1 udp 2013266431 ip private 10101 typ host
a=candidate:1 1 udp 2013266431 ip public 10101 typ host 
```
 
Is there a standard way to configure/use pion to rewrite sent candidates? Or another way to do it. While remaining host type to avoid a configuration by STUN when STUN/TURN is not used and which works as well under chrome as firefox ?

**Comment by Sean-Der** [comment, marked as answer]
Hey @mysolo 

So FireFox will not attempt to use Pion's srflx candidate if FireFox only has host candidates? I think that is worth investigation on the FireFox side, that might be hurting their connectivity success rates.

Unfortunately I don't see anything we can do without breaking public API. The quickest way to implement this on your side would de [strings.ReplaceAll](https://pkg.go.dev/strings#ReplaceAll) and rewrite `typ srflx` -> `typ host`
